src/index.tsx

```diff
@@ -28,165 +45,137 @@ interface AudioPosition {
-                  pause,
-                  stop,
-                  seek,
-                  isPlaying,
-                  mute,
-                  sound: player.current
-              }
-            : props.value
-    }, [
-        props.value,
-        playbackReady,
-        error,
-        loading,
-        loadAudio,
-        play,
-        pause,
-        stop,
-        seek,
-        mute,
-        isPlaying,
-        player
-    ])
+    useEffect(() => {
+        // unload the player on unmount
+        return () => {
+            if (playerRef.current) playerRef.current.unload()
+        }
+    }, [])
+
+    const contextValue: AudioPlayer = value
+        ? value
+        : {
+              player,
+              load,
+              error,
+              loading,
+              playing,
+              stopped,
+              ready: !loading && !error,
+          }
 
     return (
         <AudioPlayerContext.Provider value={contextValue}>
-            {props.children}
+            {children}
         </AudioPlayerContext.Provider>
     )
 }
 
-// gives programmer access to the audio context
-export const useAudioPlayer = (url?: string): AudioPlayer => {
-    const value = React.useContext(AudioPlayerContext)
+export const useAudioPlayer = (props?: AudioSrcProps): UseAudioPlayer => {
+    const { player, load, ...context } = useContext(AudioPlayerContext)!
 
-    React.useEffect(() => {
-        if (url) {
-            value.loadAudio(url)
-        }
-    }, [url, value])
+    const { src, format, autoplay } = props || {}
+
+    useEffect(() => {
+        // if useAudioPlayer is called without arguments
+        // don't do anything: the user will have access
+        // to the current context
+        if (!src) return
+        load({ src, format, autoplay })
+    }, [src, format, autoplay, load])
 
-    return value
+    return {
+        ...context,
+        play: player ? player.play.bind(player) : noop,
+        pause: player ? player.pause.bind(player) : noop,
+        stop: player ? player.stop.bind(player) : noop,
+        mute: player ? player.mute.bind(player) : noop,
+        seek: player ? player.seek.bind(player) : noop,
+    }
 }
 
 // gives current audio position state updates in an animation frame loop for animating visualizations
 export const useAudioPosition = (): AudioPosition => {
-    const { sound, playbackReady } = useAudioPlayer()
-    const [position, setPosition] = React.useState(0)
-    const [duration, setDuration] = React.useState(0)
-    const frameRequest = React.useRef<number | undefined>()
-
-    const animate = React.useCallback(() => {
-        if (playbackReady && sound.playing()) {
-            setPosition(sound.seek() as number)
-            frameRequest.current = requestAnimationFrame(animate)
-        } else {
-            cancelAnimationFrame(frameRequest.current)
-        }
-    }, [playbackReady, sound])
+    const { player, stopped, playing } = useContext(AudioPlayerContext)!
```

Comment:
@dbismut, what does the `!` operator do here at the end of line?

Reply:
The context can be null (not in real world usage but according to types). The `!` typescript operator indicates the compiler we are certain that the value exists.

Reply:
ahh cool! Didn't know about that TS feature